janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java

```diff
@@ -0,0 +1,243 @@
+        } else {
+            return rawData.length - offsetIndex[index];
+        }
+    }
+
+    private int getEntryEndOffset(final int index) {
+        if (index < offsetIndex.length - 1) {
+            return offsetIndex[index + 1];
+        } else {
+            return rawData.length;
+        }
+    }
+
+    public Entry get(final int index) {
+        final int entryBufLen = getEntryLength(index);
+
+        final ByteBuffer entryBuffer = ByteBuffer.wrap(rawData, offsetIndex[index], entryBufLen);
+
+        final byte valPos = entryBuffer.get();
+
+        final byte[] col = new byte[valPos];
+        entryBuffer.get(col);
+
+        final byte[] val = new byte[entryBufLen - valPos - 1];
+        entryBuffer.get(val);
+
+        return StaticArrayEntry.of(StaticArrayBuffer.of(col), StaticArrayBuffer.of(val));
+    }
+
+    public Entry getNoCopy(final int index) {
+        return new StaticArrayEntry(rawData, offsetIndex[index] + 1, getEntryEndOffset(index), rawData[offsetIndex[index]]);
+    }
+
+    public int numEntries() {
+        return offsetIndex.length;
+    }
+
+    public int byteSize() {
+        return rawData.length + offsetIndex.length * Integer.BYTES + 16;
+    }
+
+    public static List<BufferPage> merge(List<BufferPage> pagesToMerge, int maxPageSize) {
+        if (pagesToMerge == null || pagesToMerge.size() < 1) {
+            return pagesToMerge;
+        }
+
+        //NOTE: instead of constructing all Entries and then rebuilding offsetIndex and rawData,
+        // it could use arraycopy to copy parts of existing buffers into new ones
+        // it is possible to do in one pass, but rather complex as we'd need to keep track of current offset in new buffer,
+        // current page and offset within that page etc. In practice, this is required seldom enough to create any noticeable deficiencies.
+
+        // Just extracts all Entries from all fragmented pages into a single list, then create new full pages
+        int totalCount = 0;
+        for (BufferPage p : pagesToMerge) {
+            totalCount += p.numEntries();
+        }
+
+        Entry[] newdata = new Entry[totalCount];
+
+        totalCount = 0;
+        for (BufferPage p : pagesToMerge) {
+            for (int i = 0; i < p.numEntries(); i++) {
+                newdata[totalCount++] = p.getNoCopy(i);
+            }
+        }
+
+        int numNewPages = newdata.length / maxPageSize;
+        if (newdata.length % maxPageSize > 0) {
+            numNewPages++;
+        }
+
+        List<BufferPage> newPages = new ArrayList<>(numNewPages);
+        for (int i = 0; i < numNewPages; i++) {
+            newPages.add(BufferPageUtils.buildFromEntryArray(newdata, i * maxPageSize, Math.min(newdata.length, (i + 1) * maxPageSize)));
+        }
+
+        return newPages;
+    }
+
+    public List<BufferPage> merge(Entry[] add, int iaddp, int addLimit, Entry[] del, int idelp, int delLimit, int maxPageSize) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'merge(Entry, int, int, Entry, int, int, int)' has an NPath complexity of 728](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)